[Bug]: Panic in snapshots restore soak test for referral codes

### Problem encountered

Observed 2 panics in snapshot soak tests for referrals.

Panic 1:

```
panic: runtime error: index out of range [-1] [recovered]
	panic: runtime error: index out of range [-1]

goroutine 65 [running]:
code.vegaprotocol.io/vega/cmd/vega/commands/node.(*Command).Run.func1.1()
	/jenkins/workspace/common/system-tests-wrapper/vega/cmd/vega/commands/node/node.go:154 +0x59
panic({0x4f40da0, 0xc00032ef48})
	/usr/local/go/src/runtime/panic.go:884 +0x213
code.vegaprotocol.io/vega/core/referral.(*runningVolumes).Add(0xc00155cdf0, 0xf, {0xc002526c80, 0x40}, 0xc00013a590?)
	/jenkins/workspace/common/system-tests-wrapper/vega/core/referral/notional_volumes.go:51 +0x3fa
code.vegaprotocol.io/vega/core/referral.(*Engine).computeReferralSetsStats(0xc0006ae420, {0x5a7e558, 0xc00252f620}, {0xf, {0x213a1b78, 0xedc93b63c, 0x0}, {0x213a1b78, 0xedc93b641, 0x0}, ...})
	/jenkins/workspace/common/system-tests-wrapper/vega/core/referral/engine.go:458 +0x289
code.vegaprotocol.io/vega/core/referral.(*Engine).OnEpoch(0xc000f32a80?, {0x5a7e558?, 0xc00252f620?}, {0xf, {0x213a1b78, 0xedc93b63c, 0x0}, {0x213a1b78, 0xedc93b641, 0x0}, ...})
	/jenkins/workspace/common/system-tests-wrapper/vega/core/referral/engine.go:313 +0x78
code.vegaprotocol.io/vega/core/epochtime.(*Svc).notify(0xc0014bf7a0, {0x5a7e558, 0xc00252f620}, {0xf, {0x213a1b78, 0xedc93b63c, 0x0}, {0x213a1b78, 0xedc93b641, 0x0}, ...})
	/jenkins/workspace/common/system-tests-wrapper/vega/core/epochtime/service.go:201 +0xe5
code.vegaprotocol.io/vega/core/epochtime.(*Svc).OnTick(0xc0014bf7a0, {0x5a7e558, 0xc00252f620}, {0x1?, 0x5a993b0?, 0x0?})
	/jenkins/workspace/common/system-tests-wrapper/vega/core/epochtime/service.go:131 +0x718
code.vegaprotocol.io/vega/core/vegatime.(*Svc).notify(...)
	/jenkins/workspace/common/system-tests-wrapper/vega/core/vegatime/service.go:104
code.vegaprotocol.io/vega/core/vegatime.(*Svc).SetTimeNow(0xc000f32980, {0x5a7e558, 0xc00252f620}, {0x40?, 0x4983500?, 0x0?})
	/jenkins/workspace/common/system-tests-wrapper/vega/core/vegatime/service.go:77 +0x24f
code.vegaprotocol.io/vega/core/processor.(*App).OnBeginBlock(_, {{0xc001558ee0, 0x20, 0x20}, {{0xb, 0x1}, {0xc0015a57c0, 0xb}, 0x9d, {0x9b67ac7, ...}, ...}, ...})
	/jenkins/workspace/common/system-tests-wrapper/vega/core/processor/abci.go:825 +0x747
code.vegaprotocol.io/vega/core/blockchain/abci.(*App).BeginBlock(_, {{0xc001558ee0, 0x20, 0x20}, {{0xb, 0x1}, {0xc0015a57c0, 0xb}, 0x9d, {0x9b67ac7, ...}, ...}, ...})
	/jenkins/workspace/common/system-tests-wrapper/vega/core/blockchain/abci/abci.go:46 +0x75
code.vegaprotocol.io/vega/cmd/vega/commands/node.(*appW).BeginBlock(_, {{0xc001558ee0, 0x20, 0x20}, {{0xb, 0x1}, {0xc0015a57c0, 0xb}, 0x9d, {0x9b67ac7, ...}, ...}, ...})
	/jenkins/workspace/common/system-tests-wrapper/vega/cmd/vega/commands/node/app_wrapper.go:64 +0x55
github.com/tendermint/tendermint/abci/client.(*localClient).BeginBlockSync(_, {{0xc001558ee0, 0x20, 0x20}, {{0xb, 0x1}, {0xc0015a57c0, 0xb}, 0x9d, {0x9b67ac7, ...}, ...}, ...})
	/jenkins/GOPATH/pkg/mod/github.com/vegaprotocol/cometbft@v0.34.28-0.20230322133204-3d8588de736e/abci/client/local_client.go:280 +0x118
github.com/tendermint/tendermint/proxy.(*appConnConsensus).BeginBlockSync(_, {{0xc001558ee0, 0x20, 0x20}, {{0xb, 0x1}, {0xc0015a57c0, 0xb}, 0x9d, {0x9b67ac7, ...}, ...}, ...})
	/jenkins/GOPATH/pkg/mod/github.com/vegaprotocol/cometbft@v0.34.28-0.20230322133204-3d8588de736e/proxy/app_conn.go:81 +0x55
github.com/tendermint/tendermint/state.execBlockOnProxyApp({0x5a7c450?, 0xc0016e63d8}, {0x5a90460, 0xc001bbaa40}, 0xc00247fa40, {0x5aa1da0, 0xc0002420d8}, 0xc00076b6c0?)
	/jenkins/GOPATH/pkg/mod/github.com/vegaprotocol/cometbft@v0.34.28-0.20230322133204-3d8588de736e/state/execution.go:307 +0x51d
github.com/tendermint/tendermint/state.ExecCommitBlock({0x5a90460, 0xc001bbaa40}, 0xc00247fa40, {0x5a7c450, 0xc0016e63d8}, {0x5aa1da0?, 0xc0002420d8?}, 0xc000124060?)
	/jenkins/GOPATH/pkg/mod/github.com/vegaprotocol/cometbft@v0.34.28-0.20230322133204-3d8588de736e/state/execution.go:537 +0x67
github.com/tendermint/tendermint/consensus.(*Handshaker).replayBlocks(_, {{{0xb, 0x1}, {0xc0015a4740, 0x7}}, {0xc0015a4750, 0xb}, 0x1, 0xe2, {{0xc0013b2780, ...}, ...}, ...}, ...)
	/jenkins/GOPATH/pkg/mod/github.com/vegaprotocol/cometbft@v0.34.28-0.20230322133204-3d8588de736e/consensus/replay.go:471 +0x248
github.com/tendermint/tendermint/consensus.(*Handshaker).ReplayBlocks(_, {{{0xb, 0x1}, {0xc0015a4740, 0x7}}, {0xc0015a4750, 0xb}, 0x1, 0xe2, {{0xc0013b2780, ...}, ...}, ...}, ...)
	/jenkins/GOPATH/pkg/mod/github.com/vegaprotocol/cometbft@v0.34.28-0.20230322133204-3d8588de736e/consensus/replay.go:393 +0x479
github.com/tendermint/tendermint/consensus.(*Handshaker).Handshake(0xc00013cd80, {0x5aa8ac8, 0xc002640750})
	/jenkins/GOPATH/pkg/mod/github.com/vegaprotocol/cometbft@v0.34.28-0.20230322133204-3d8588de736e/consensus/replay.go:268 +0x3d4
github.com/tendermint/tendermint/node.doHandshake({_, _}, {{{0xb, 0x1}, {0xc0015a4740, 0x7}}, {0xc0015a4750, 0xb}, 0x1, 0xe2, ...}, ...)
	/jenkins/GOPATH/pkg/mod/github.com/vegaprotocol/cometbft@v0.34.28-0.20230322133204-3d8588de736e/node/node.go:329 +0x1b8
github.com/tendermint/tendermint/node.NewNode(0xc00014cb40, {0x5a76ba0, 0xc00075a320}, 0xc00077abc0, {0x5a59fc0, 0xc0015a2f00}, 0x206e412220202020?, 0x6c616e6f6974706f?, 0xc00077ac00, {0x5a7c450, ...}, ...)
	/jenkins/GOPATH/pkg/mod/github.com/vegaprotocol/cometbft@v0.34.28-0.20230322133204-3d8588de736e/node/node.go:779 +0x597
code.vegaprotocol.io/vega/core/blockchain/abci.NewTmNode({{0x0}, 0x1, 0x1, 0x0, 0x0, {0xc000078c10, 0xa}, {{0x0}, {0xc00062e558, 0x15}}, ...}, ...)
	/jenkins/workspace/common/system-tests-wrapper/vega/core/blockchain/abci/tm_node.go:83 +0x725
code.vegaprotocol.io/vega/cmd/vega/commands/node.(*Command).startABCI(0xc0010ae900, 0x0?, {0x5aa25a0, 0xc000624360}, {0x7ffe449cd0e6, 0x53}, {0x0?, 0x0?}, {0x0, 0x0})
	/jenkins/workspace/common/system-tests-wrapper/vega/cmd/vega/commands/node/node.go:426 +0x1b6
code.vegaprotocol.io/vega/cmd/vega/commands/node.(*Command).startBlockchain(0xc0010ae900, 0x0?, {0x7ffe449cd0e6?, 0x0?}, {0x0, 0x0}, {0x0, 0x0})
	/jenkins/workspace/common/system-tests-wrapper/vega/cmd/vega/commands/node/node.go:313 +0x3ab
code.vegaprotocol.io/vega/cmd/vega/commands/node.(*Command).Run.func1()
	/jenkins/workspace/common/system-tests-wrapper/vega/cmd/vega/commands/node/node.go:157 +0x99
created by code.vegaprotocol.io/vega/cmd/vega/commands/node.(*Command).Run
	/jenkins/workspace/common/system-tests-wrapper/vega/cmd/vega/commands/node/node.go:148 +0x74f
```

Panic 2:

```
panic: block.AppHash does not match AppHash after replay. Got 1A1FEE92D6B41A25A96DE5ED76C8E6C6CF5A8033718894AF97D0BD7F0D94EE0B, expected 3E3670820258EDCF99B58B19AAE952A515E4798324DFBEDAF81CE69D83BD9F76.

Block: Block{
  Header{
    Version:        {11 1}
    ChainID:        testnet-001
    Height:         199
    Time:           2023-09-13 13:55:41.943637674 +0000 UTC
    LastBlockID:    45F9E6BC70734858BFE3AD81ADF8BFCC16E9584E7E0EDC65E28610474532365D:1:41F392470208
    LastCommit:     23E8EC695AB0E9BF693A96B6C5008FAFEBB2F31BE6287474DFEBE16067858230
    Data:           E3B0C44298FC1C149AFBF4C8996FB92427AE41E4649B934CA495991B7852B855
    Validators:     9F7D347BC237DEA341455A57DB079EEC33CA94733643A993F9489DE1B7CDFDC0
    NextValidators: 9F7D347BC237DEA341455A57DB079EEC33CA94733643A993F9489DE1B7CDFDC0
    App:            3E3670820258EDCF99B58B19AAE952A515E4798324DFBEDAF81CE69D83BD9F76
    Consensus:      C8CD39A92B22597AC33565A59990B4D1E440B15621B5E932D1D5B39A8F45E9DB
    Results:        E3B0C44298FC1C149AFBF4C8996FB92427AE41E4649B934CA495991B7852B855
    Evidence:       E3B0C44298FC1C149AFBF4C8996FB92427AE41E4649B934CA495991B7852B855
    Proposer:       4CE183A787568DAA2101593576C4B19F91B4EF30
  }#C7BF3E57B46BEC893421E4BBE75E25CA8556569158E3ED810B68F72A77566596
  Data{
    
  }#E3B0C44298FC1C149AFBF4C8996FB92427AE41E4649B934CA495991B7852B855
  EvidenceData{
    
  }#E3B0C44298FC1C149AFBF4C8996FB92427AE41E4649B934CA495991B7852B855
  Commit{
    Height:     198
    Round:      0
    BlockID:    45F9E6BC70734858BFE3AD81ADF8BFCC16E9584E7E0EDC65E28610474532365D:1:41F392470208
    Signatures:
      CommitSig{76241F331882 by 4CE183A78756 on 2 @ 2023-09-13T13:55:41.943637674Z}
      CommitSig{2BFE34E2F57C by CA0BB2F7D7FB on 2 @ 2023-09-13T13:55:42.043458186Z}
  }#23E8EC695AB0E9BF693A96B6C5008FAFEBB2F31BE6287474DFEBE16067858230
}#C7BF3E57B46BEC893421E4BBE75E25CA8556569158E3ED810B68F72A77566596
 [recovered]
	panic: block.AppHash does not match AppHash after replay. Got 1A1FEE92D6B41A25A96DE5ED76C8E6C6CF5A8033718894AF97D0BD7F0D94EE0B, expected 3E3670820258EDCF99B58B19AAE952A515E4798324DFBEDAF81CE69D83BD9F76.

Block: Block{
  Header{
    Version:        {11 1}
    ChainID:        testnet-001
    Height:         199
    Time:           2023-09-13 13:55:41.943637674 +0000 UTC
    LastBlockID:    45F9E6BC70734858BFE3AD81ADF8BFCC16E9584E7E0EDC65E28610474532365D:1:41F392470208
    LastCommit:     23E8EC695AB0E9BF693A96B6C5008FAFEBB2F31BE6287474DFEBE16067858230
    Data:           E3B0C44298FC1C149AFBF4C8996FB92427AE41E4649B934CA495991B7852B855
    Validators:     9F7D347BC237DEA341455A57DB079EEC33CA94733643A993F9489DE1B7CDFDC0
    NextValidators: 9F7D347BC237DEA341455A57DB079EEC33CA94733643A993F9489DE1B7CDFDC0
    App:            3E3670820258EDCF99B58B19AAE952A515E4798324DFBEDAF81CE69D83BD9F76
    Consensus:      C8CD39A92B22597AC33565A59990B4D1E440B15621B5E932D1D5B39A8F45E9DB
    Results:        E3B0C44298FC1C149AFBF4C8996FB92427AE41E4649B934CA495991B7852B855
    Evidence:       E3B0C44298FC1C149AFBF4C8996FB92427AE41E4649B934CA495991B7852B855
    Proposer:       4CE183A787568DAA2101593576C4B19F91B4EF30
  }#C7BF3E57B46BEC893421E4BBE75E25CA8556569158E3ED810B68F72A77566596
  Data{
    
  }#E3B0C44298FC1C149AFBF4C8996FB92427AE41E4649B934CA495991B7852B855
  EvidenceData{
    
  }#E3B0C44298FC1C149AFBF4C8996FB92427AE41E4649B934CA495991B7852B855
  Commit{
    Height:     198
    Round:      0
    BlockID:    45F9E6BC70734858BFE3AD81ADF8BFCC16E9584E7E0EDC65E28610474532365D:1:41F392470208
    Signatures:
      CommitSig{76241F331882 by 4CE183A78756 on 2 @ 2023-09-13T13:55:41.943637674Z}
      CommitSig{2BFE34E2F57C by CA0BB2F7D7FB on 2 @ 2023-09-13T13:55:42.043458186Z}
  }#23E8EC695AB0E9BF693A96B6C5008FAFEBB2F31BE6287474DFEBE16067858230
}#C7BF3E57B46BEC893421E4BBE75E25CA8556569158E3ED810B68F72A77566596


goroutine 27 [running]:
code.vegaprotocol.io/vega/cmd/vega/commands/node.(*Command).Run.func1.1()
	/jenkins/workspace/common/system-tests-wrapper/vega/cmd/vega/commands/node/node.go:154 +0x59
panic({0x49843c0, 0xc0014621d0})
	/usr/local/go/src/runtime/panic.go:884 +0x213
github.com/tendermint/tendermint/consensus.assertAppHashEqualsOneFromBlock({0xc0040ba660, 0x20, 0x20}, 0xc0040c6000)
	/jenkins/GOPATH/pkg/mod/github.com/vegaprotocol/cometbft@v0.34.28-0.20230322133204-3d8588de736e/consensus/replay.go:515 +0x10d
github.com/tendermint/tendermint/consensus.(*Handshaker).replayBlocks(_, {{{0xb, 0x1}, {0xc0016f25c0, 0x7}}, {0xc0016f26f0, 0xb}, 0x1, 0xe2, {{0xc0013ec4e0, ...}, ...}, ...}, ...)
	/jenkins/GOPATH/pkg/mod/github.com/vegaprotocol/cometbft@v0.34.28-0.20230322133204-3d8588de736e/consensus/replay.go:468 +0x1fd
github.com/tendermint/tendermint/consensus.(*Handshaker).ReplayBlocks(_, {{{0xb, 0x1}, {0xc0016f25c0, 0x7}}, {0xc0016f26f0, 0xb}, 0x1, 0xe2, {{0xc0013ec4e0, ...}, ...}, ...}, ...)
	/jenkins/GOPATH/pkg/mod/github.com/vegaprotocol/cometbft@v0.34.28-0.20230322133204-3d8588de736e/consensus/replay.go:393 +0x479
github.com/tendermint/tendermint/consensus.(*Handshaker).Handshake(0xc000178d80, {0x5aa8ac8, 0xc000f6c340})
	/jenkins/GOPATH/pkg/mod/github.com/vegaprotocol/cometbft@v0.34.28-0.20230322133204-3d8588de736e/consensus/replay.go:268 +0x3d4
github.com/tendermint/tendermint/node.doHandshake({_, _}, {{{0xb, 0x1}, {0xc0016f25c0, 0x7}}, {0xc0016f26f0, 0xb}, 0x1, 0xe2, ...}, ...)
	/jenkins/GOPATH/pkg/mod/github.com/vegaprotocol/cometbft@v0.34.28-0.20230322133204-3d8588de736e/node/node.go:329 +0x1b8
github.com/tendermint/tendermint/node.NewNode(0xc0013b4a00, {0x5a76ba0, 0xc0002b74a0}, 0xc002bb5760, {0x5a59fc0, 0xc002bb8b28}, 0x0?, 0x0?, 0xc002bb57a0, {0x5a7c450, ...}, ...)
	/jenkins/GOPATH/pkg/mod/github.com/vegaprotocol/cometbft@v0.34.28-0.20230322133204-3d8588de736e/node/node.go:779 +0x597
code.vegaprotocol.io/vega/core/blockchain/abci.NewTmNode({{0x0}, 0x1, 0x1, 0x0, 0x0, {0xc000edf550, 0xa}, {{0x0}, {0xc0005ce510, 0x15}}, ...}, ...)
	/jenkins/workspace/common/system-tests-wrapper/vega/core/blockchain/abci/tm_node.go:83 +0x725
code.vegaprotocol.io/vega/cmd/vega/commands/node.(*Command).startABCI(0xc00108e900, 0x2a9e7c0?, {0x5aa25a0, 0xc0015eff40}, {0x7fffcf1580e6, 0x53}, {0x0?, 0xc002662d80?}, {0x0, 0x0})
	/jenkins/workspace/common/system-tests-wrapper/vega/cmd/vega/commands/node/node.go:426 +0x1b6
code.vegaprotocol.io/vega/cmd/vega/commands/node.(*Command).startBlockchain(0xc00108e900, 0x0?, {0x7fffcf1580e6?, 0x0?}, {0x0, 0x0}, {0x0, 0x0})
	/jenkins/workspace/common/system-tests-wrapper/vega/cmd/vega/commands/node/node.go:313 +0x3ab
code.vegaprotocol.io/vega/cmd/vega/commands/node.(*Command).Run.func1()
	/jenkins/workspace/common/system-tests-wrapper/vega/cmd/vega/commands/node/node.go:157 +0x99
created by code.vegaprotocol.io/vega/cmd/vega/commands/node.(*Command).Run
	/jenkins/workspace/common/system-tests-wrapper/vega/cmd/vega/commands/node/node.go:148 +0x74f
```

### Observed behaviour

Panics as mentioned above.

### Expected behaviour

No panic

### Steps to reproduce

```bash
Run the below system test:

System Test Name - test_create_referral_set_without_team
System Test Branch - feat/teams-tests
```


### Software version

vega - develop

### Failing test

test_create_referral_set_without_team

### Jenkins run

https://jenkins.vega.rocks/job/common/job/system-tests-wrapper/8434/

### Configuration used

_No response_

### Relevant log output

```shell
Soak Test Logs:

https://jenkins.vega.rocks/job/common/job/system-tests-wrapper/8434/artifact/soak-test/

Panic 1 :
https://jenkins.vega.rocks/job/common/job/system-tests-wrapper/8434/artifact/soak-test/err-node-154.log

Panic 2:
https://jenkins.vega.rocks/job/common/job/system-tests-wrapper/8434/artifact/soak-test/err-node-176.log
```

--- a/core/referral/engine.go
+++ b/core/referral/engine.go
@@ -440,7 +440,14 @@ func (e *Engine) loadReferralSetsFromSnapshot(setsProto *snapshotpb.ReferralSets
 				value: volumeNum,
 			})
 		}
-		e.referralSetsNotionalVolumes.runningVolumesBySet[setID] = runningVolumes
+
+		// set only if the running volume is not empty, or it will panic
+		// down the line when trying to add new ones.
+		// the creation of runningVolumeBySet is done in the Add method of the
+		// runningVolumes type.
+		if len(runningVolumes) > 0 {
+			e.referralSetsNotionalVolumes.runningVolumesBySet[setID] = runningVolumes
+		}
 
 		e.sets[setID] = newSet
 	}

